In [166]:
import pandas as pd
import numpy as np
import re

data_frame.apply(lambda data_frame: name_function( data_frame['col1'], data_frame['col2']), axis=1 )

-->itera cada columna con la función que le des


    def new_ranking3(genre,rating):

        if "Thriller" in genre:
        
            return rating+1
            
        elif "Comedy" in genre:
        
            return rating+1
            
        else:
        
            return rating+0
            
            
data_set["new_ranking3"] = data_set.apply(lambda data_set: new_ranking2(data_set['Genre'], data_set['Rating']), axis=1)
data_set.head(100)

In [167]:
# CHALLENGE 1
data_set = pd.read_csv("./data/input/IMDB-Movie-Data.csv")
display(data_set)

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
0                  121     8.1  757074              333.13       76.0  
1                  124     7.0  485820              126.46       65.0  
2                  117     7.3  157606              138.12       62.0  
3                  108     7.2   60545              270.32       59.0  
4                  123     6.2  393727              325.02       40.0  
..                 ...     ...     ...                 ...        ...  
995                111     6.2   27585                 NaN       45.0  
996                 94     5.5   73152               17.54       46.0  
997                 98     6.2   70699               58.01       50.0  
998                 93     5.6    4881                 NaN       22.0  
999                 87     5.3   12435               19.64       11.0  

[1000 rows x 12 columns]

From 0 to 999 ==> 1
From 1000 to 9999 ==> 2
From 10000 to 99999 ==> 3
From 100000 to 999999 ==> 4
More than 1000000 ==> 5

In [168]:
data_set_votes = data_set["Votes"]
data_set_votes

0      757074
1      485820
2      157606
3       60545
4      393727
        ...  
995     27585
996     73152
997     70699
998      4881
999     12435
Name: Votes, Length: 1000, dtype: int64

In [169]:
def create_bin(data_set_votes):
    if data_set_votes >= 0 and data_set_votes < 1000:
        return 1
    elif data_set_votes >= 1000 and data_set_votes < 10000:
        return 2
    elif data_set_votes >= 10000 and data_set_votes < 100000:
        return 3
    elif data_set_votes >= 100000 and data_set_votes < 1000000:
        return 4
    elif data_set_votes >= 1000000:
        return 5
    else:
        return "error"
    

In [170]:
bin = data_set["Votes"].apply(create_bin)
bin

0      4
1      4
2      4
3      3
4      4
      ..
995    3
996    3
997    3
998    2
999    3
Name: Votes, Length: 1000, dtype: int64

In [171]:
data_set["bin"]=data_set["Votes"].apply(create_bin)
data_set

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  
0                  121     8.1  757074              333.13       76.0    4  
1                  124     7.0  485820              126.46       65.0    4  
2                  117     7.3  157606              138.12       62.0    4  
3                  108     7.2   60545              270.32       59.0    3  
4                  123     6.2  393727              325.02       40.0    4  
..                 ...     ...     ...                 ...        ...  ...  
995                111     6.2   27585                 NaN       45.0    3  
996                 94     5.5   73152               17.54       46.0    3  
997                 98     6.2   70699               58.01       50.0    3  
998                 93     5.6    4881                 NaN       22.0    2  
999                 87     5.3   12435               19.64       11.0    3  

[1000 rows x 13 columns]

In [172]:
#Challenge 2

In [173]:
#lambda method
#def fun_div(p1,p2):
#    return p1/p2
#data_set["revenue_min"]=data_set.apply(lambda dataset: funcion(data_set["Revenue (Millions)",data_set["Runtime (Minutes)]), axis=1)

In [174]:
data_set["revenue_min"]=data_set["Revenue (Millions)"]/data_set["Runtime (Minutes)"]
data_set

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  \
0                  121     8.1  757074              333.13       76.0    4   
1                  124     7.0  485820              126.46       65.0    4   
2                  117     7.3  157606              138.12       62.0    4   
3                  108     7.2   60545              270.32       59.0    3   
4                  123     6.2  393727              325.02       40.0    4   
..                 ...     ...     ...                 ...        ...  ...   
995                111     6.2   27585                 NaN       45.0    3   
996                 94     5.5   73152               17.54       46.0    3   
997                 98     6.2   70699               58.01       50.0    3   
998                 93     5.6    4881                 NaN       22.0    2   
999                 87     5.3   12435               19.64       11.0    3   

     revenue_min  
0       2.753140  
1       1.019839  
2       1.180513  
3       2.502963  
4       2.642439  
..           ...  
995          NaN  
996     0.186596  
997     0.591939  
998          NaN  
999     0.225747  

[1000 rows x 14 columns]

In [175]:
#Challenge 3 We want to create a new ranking where we add 1 point if the genre is thriller but subtract 1 point if the genre is comedy.

In [176]:
data_set_genre = data_set["Genre"]
data_set_rating = data_set["Rating"]

def new_ranking(data_set_rating,data_set_genre):
    lst=[]
    for i,e in zip(data_set_rating,data_set_genre):
        if "Thriller" in e:
            lst.append(i+1)
        elif "Comedy" in e:
            lst.append(i-1)
        else:
            lst.append(i)
    return lst

x = new_ranking(data_set_rating,data_set_genre)


In [177]:

def new_ranking2(genre,rating):
        if re.match(".*Thriller.*",genre):
            return rating+1
        elif re.match(".*Comedy.*",genre):
            return rating+1
        else:
            return rating+0

In [178]:
data_set["new_ranking2"] = data_set.apply(lambda data_set: new_ranking2(data_set['Genre'], data_set['Rating']), axis=1)
data_set.head(100)

Rank                    Title                     Genre  \
0      1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1      2               Prometheus  Adventure,Mystery,Sci-Fi   
2      3                    Split           Horror,Thriller   
3      4                     Sing   Animation,Comedy,Family   
4      5            Suicide Squad  Action,Adventure,Fantasy   
..   ...                      ...                       ...   
95    96            The Nice Guys       Action,Comedy,Crime   
96    97            Kimi no na wa   Animation,Drama,Fantasy   
97    98                 The Void     Horror,Mystery,Sci-Fi   
98    99         Personal Shopper    Drama,Mystery,Thriller   
99   100             The Departed      Crime,Drama,Thriller   

                                          Description              Director  \
0   A group of intergalactic criminals are forced ...            James Gunn   
1   Following clues to the origin of mankind, a te...          Ridley Scott   
2   Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3   In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4   A secret government agency recruits some of th...            David Ayer   
..                                                ...                   ...   
95  In 1970s Los Angeles, a mismatched pair of pri...           Shane Black   
96  Two strangers find themselves linked in a biza...        Makoto Shinkai   
97  Shortly after delivering a patient to an under...      Jeremy Gillespie   
98  A personal shopper in Paris refuses to leave t...       Olivier Assayas   
99  An undercover cop and a mole in the police att...       Martin Scorsese   

                                               Actors  Year  \
0   Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1   Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2   James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3   Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4   Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                ...   ...   
95  Russell Crowe, Ryan Gosling, Angourie Rice, Ma...  2016   
96  Ryûnosuke Kamiki, Mone Kamishiraishi, Ryô Nari...  2016   
97  Aaron Poole, Kenneth Welsh,Daniel Fathers, Kat...  2016   
98  Kristen Stewart, Lars Eidinger, Sigrid Bouaziz...  2016   
99  Leonardo DiCaprio, Matt Damon, Jack Nicholson,...  2006   

    Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  \
0                 121     8.1  757074              333.13       76.0    4   
1                 124     7.0  485820              126.46       65.0    4   
2                 117     7.3  157606              138.12       62.0    4   
3                 108     7.2   60545              270.32       59.0    3   
4                 123     6.2  393727              325.02       40.0    4   
..                ...     ...     ...                 ...        ...  ...   
95                116     7.4  175067               36.25       70.0    4   
96                106     8.6   34110                4.68       79.0    3   
97                 90     5.8    9247                0.15       62.0    2   
98                105     6.3   10181                1.29       77.0    3   
99                151     8.5  937414              132.37       85.0    4   

    revenue_min  new_ranking2  
0      2.753140           8.1  
1      1.019839           7.0  
2      1.180513           8.3  
3      2.502963           8.2  
4      2.642439           6.2  
..          ...           ...  
95     0.312500           8.4  
96     0.044151           8.6  
97     0.001667           5.8  
98     0.012286           7.3  
99     0.876623           9.5  

[100 rows x 15 columns]

In [179]:
def new_ranking3(genre,rating):
        if "Thriller" in genre:
            return rating+1
        elif "Comedy" in genre:
            return rating+1
        else:
            return rating+0
        
data_set["new_ranking3"] = data_set.apply(lambda data_set: new_ranking2(data_set['Genre'], data_set['Rating']), axis=1)
data_set.head(100)

Rank                    Title                     Genre  \
0      1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1      2               Prometheus  Adventure,Mystery,Sci-Fi   
2      3                    Split           Horror,Thriller   
3      4                     Sing   Animation,Comedy,Family   
4      5            Suicide Squad  Action,Adventure,Fantasy   
..   ...                      ...                       ...   
95    96            The Nice Guys       Action,Comedy,Crime   
96    97            Kimi no na wa   Animation,Drama,Fantasy   
97    98                 The Void     Horror,Mystery,Sci-Fi   
98    99         Personal Shopper    Drama,Mystery,Thriller   
99   100             The Departed      Crime,Drama,Thriller   

                                          Description              Director  \
0   A group of intergalactic criminals are forced ...            James Gunn   
1   Following clues to the origin of mankind, a te...          Ridley Scott   
2   Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3   In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4   A secret government agency recruits some of th...            David Ayer   
..                                                ...                   ...   
95  In 1970s Los Angeles, a mismatched pair of pri...           Shane Black   
96  Two strangers find themselves linked in a biza...        Makoto Shinkai   
97  Shortly after delivering a patient to an under...      Jeremy Gillespie   
98  A personal shopper in Paris refuses to leave t...       Olivier Assayas   
99  An undercover cop and a mole in the police att...       Martin Scorsese   

                                               Actors  Year  \
0   Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1   Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2   James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3   Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4   Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                ...   ...   
95  Russell Crowe, Ryan Gosling, Angourie Rice, Ma...  2016   
96  Ryûnosuke Kamiki, Mone Kamishiraishi, Ryô Nari...  2016   
97  Aaron Poole, Kenneth Welsh,Daniel Fathers, Kat...  2016   
98  Kristen Stewart, Lars Eidinger, Sigrid Bouaziz...  2016   
99  Leonardo DiCaprio, Matt Damon, Jack Nicholson,...  2006   

    Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  \
0                 121     8.1  757074              333.13       76.0    4   
1                 124     7.0  485820              126.46       65.0    4   
2                 117     7.3  157606              138.12       62.0    4   
3                 108     7.2   60545              270.32       59.0    3   
4                 123     6.2  393727              325.02       40.0    4   
..                ...     ...     ...                 ...        ...  ...   
95                116     7.4  175067               36.25       70.0    4   
96                106     8.6   34110                4.68       79.0    3   
97                 90     5.8    9247                0.15       62.0    2   
98                105     6.3   10181                1.29       77.0    3   
99                151     8.5  937414              132.37       85.0    4   

    revenue_min  new_ranking2  new_ranking3  
0      2.753140           8.1           8.1  
1      1.019839           7.0           7.0  
2      1.180513           8.3           8.3  
3      2.502963           8.2           8.2  
4      2.642439           6.2           6.2  
..          ...           ...           ...  
95     0.312500           8.4           8.4  
96     0.044151           8.6           8.6  
97     0.001667           5.8           5.8  
98     0.012286           7.3           7.3  
99     0.876623           9.5           9.5  

[100 rows x 16 columns]

CHALLENGE 4:
We want to know if the sum of the ASCII value of every character of the movie title divided by the number of votes retrieve a prime number...remember, prime numbers are integers

In [180]:
def prime(title, votes):
    sum_title = 0
    for i in title:
        sum_title +=ord(i)
    if sum_title%votes == 0:
        return "prime"
    else:
        return "no prime"

In [181]:
data_set["prime"] = data_set.apply(lambda data_set: prime(data_set['Title'], data_set['Votes']), axis=1)
data_set

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  \
0                  121     8.1  757074              333.13       76.0    4   
1                  124     7.0  485820              126.46       65.0    4   
2                  117     7.3  157606              138.12       62.0    4   
3                  108     7.2   60545              270.32       59.0    3   
4                  123     6.2  393727              325.02       40.0    4   
..                 ...     ...     ...                 ...        ...  ...   
995                111     6.2   27585                 NaN       45.0    3   
996                 94     5.5   73152               17.54       46.0    3   
997                 98     6.2   70699               58.01       50.0    3   
998                 93     5.6    4881                 NaN       22.0    2   
999                 87     5.3   12435               19.64       11.0    3   

     revenue_min  new_ranking2  new_ranking3     prime  
0       2.753140           8.1           8.1  no prime  
1       1.019839           7.0           7.0  no prime  
2       1.180513           8.3           8.3  no prime  
3       2.502963           8.2           8.2  no prime  
4       2.642439           6.2           6.2  no prime  
..           ...           ...           ...       ...  
995          NaN           6.2           6.2  no prime  
996     0.186596           5.5           5.5  no prime  
997     0.591939           6.2           6.2  no prime  

CHALLENGE 5:
Feel free to propose your own ranking based in aggregations of at least 3 columns of the dataset.

In [182]:
data_set["Votes"].max()

1791916

In [188]:
#(Rating*10 + Metascore)/2

def alvaro_ranking(ranking, metascore, votes):
    final_ranking = (ranking*10 + metascore)/2
    if votes >=0 and votes <50000:
        return f"{final_ranking} - poorly voted"
    elif votes >=50000 and votes <100000:
        return f"{final_ranking} - moderately voted" 
    elif votes >=100000:
        return f"{final_ranking} - hihgly voted"    

In [189]:
data_set["alvaro_ranking"] = data_set.apply(lambda data_set: alvaro_ranking(data_set['Rating'], data_set['Metascore'], data_set['Votes']), axis=1)
data_set

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  \
0                  121     8.1  757074              333.13       76.0    4   
1                  124     7.0  485820              126.46       65.0    4   
2                  117     7.3  157606              138.12       62.0    4   
3                  108     7.2   60545              270.32       59.0    3   
4                  123     6.2  393727              325.02       40.0    4   
..                 ...     ...     ...                 ...        ...  ...   
995                111     6.2   27585                 NaN       45.0    3   
996                 94     5.5   73152               17.54       46.0    3   
997                 98     6.2   70699               58.01       50.0    3   
998                 93     5.6    4881                 NaN       22.0    2   
999                 87     5.3   12435               19.64       11.0    3   

     revenue_min  new_ranking2  new_ranking3     prime  \
0       2.753140           8.1           8.1  no prime   
1       1.019839           7.0           7.0  no prime   
2       1.180513           8.3           8.3  no prime   
3       2.502963           8.2           8.2  no prime   
4       2.642439           6.2           6.2  no prime   
..           ...           ...           ...       ...   
995          NaN           6.2           6.2  no prime   
996     0.186596           5.5           5.5  no prime   
997     0.591939           6.2           6.2  n

CHALLENGE 6:
We want to know which movies might have hidden paterns in their description. A way to know that is finding those movies which the sum of all numeric values of the string description hash (SHA256) are between their revenue and their number of votes.

In [227]:
#hashlib.sha256()
def sha256(description, revenue, votes):
    enconde = description.encode()
    sha256 = hashlib.sha256(enconde).hexdigest()
    sha256_nums = re.sub("[a-zA-Z]","",sha256)
    sha256_nums_map = list(map(int, sha256_nums))
    sha256_nums_map_sum = sum(sha256_nums_map)
    if sha256_nums_map_sum < revenue*1000 and sha256_nums_map_sum > votes:
        return "hidden paterns"
    else:
        return "No hidden paterns"

In [228]:
data_set["hidden paterns"] = data_set.apply(lambda data_set: sha256(data_set['Description'], data_set['Revenue (Millions)'], data_set['Votes']), axis=1)
data_set

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  bin  \
0                  121     8.1  757074              333.13       76.0    4   
1                  124     7.0  485820              126.46       65.0    4   
2                  117     7.3  157606              138.12       62.0    4   
3                  108     7.2   60545              270.32       59.0    3   
4                  123     6.2  393727              325.02       40.0    4   
..                 ...     ...     ...                 ...        ...  ...   
995                111     6.2   27585                 NaN       45.0    3   
996                 94     5.5   73152               17.54       46.0    3   
997                 98     6.2   70699               58.01       50.0    3   
998                 93     5.6    4881                 NaN       22.0    2   
999                 87     5.3   12435               19.64       11.0    3   

     revenue_min  new_ranking2  new_ranking3     prime  \
0       2.753140           8.1           8.1  no prime   
1       1.019839           7.0           7.0  no prime   
2       1.180513           8.3           8.3  no prime   
3       2.502963           8.2           8.2  no prime   
4       2.642439           6.2           6.2  no prime   
..           ...           ...           ...       ...   
995          NaN           6.2           6.2  no prime   
996     0.186596           5.5           5.5  no prime   
997     0.591939           6.2           6.2  n